src/OrchardCore.Modules/OrchardCore.Menu/Controllers/AdminController.cs

```diff
@@ -270,16 +270,18 @@ public async Task<IActionResult> Delete(string menuContentItemId, string menuIte
                 return NotFound();
             }
 
+            var menuContentAsJson = (JsonObject)menu.Content;
             // Look for the target menu item in the hierarchy.
-            var menuItem = FindMenuItem((JsonObject)menu.Content, menuItemId);
+            var menuItem = FindMenuItem(menuContentAsJson, menuItemId);
 
             // Couldn't find targeted menu item.
             if (menuItem == null)
             {
                 return NotFound();
             }
 
-            menu.Content.Remove(menuItemId);
+            var menuItems = menuContentAsJson[nameof(MenuItemsListPart)]["MenuItems"] as JsonArray;
+            menuItems.Remove(menuItem);
```

Comment:
Okay, I may have another problem

Reply:
@sarahelsaig updated , at  https://github.com/OrchardCMS/OrchardCore/pull/15509

Reply:
> Sorry for the nitpick, but why did you use `as JsonArray` here? Is it possible that this won't be here or won't be the right type? Because right now it will just cause an NRE on the next line in that scenario. I suggest either doing null-forgiving on the Remove as well (`menuItems?.Remove(menuItem);`) or throwing an explicit exception.


Because there was a logical error in the last change here

Reply:
Could you elaborate?

Reply:
Sorry, my English is not good. You can debug the original code, which attempts to remove a menuId directly from a JObject

Reply:
>Sorry for the nitpick, but why did you use as JsonArray here? Is it possible that this won't be here or won't be the right type? Because right now it will just cause an NRE on the next line in that scenario. I suggest either doing null-forgiving on the Remove as well (menuItems?.Remove(menuItem);) or throwing an explicit exception.

The `MenuItems` is always an array

Reply:
> The MenuItems is always an array

Thanks for the clarification.

> You can debug the original code, which attempts to remove a menuId directly from a jsonObject

I've tried understanding `JsonDynamicObject`. It's the same as the inner `_jsonObject`, except it also has a `_dictionary`. This `_dictionary` field seems like a cache to me. So this could be fixed by emptying it after the json object is updated.

Reply:
@hyzx86 I've fixed it using the additional "Remove" methods from your branch. Please check out my PR (https://github.com/hyzx86/OrchardCore/pull/1003).